src/libponyrt/ds/hash.c

```diff
@@ -22,13 +22,18 @@ static void* search(hashmap_t* map, size_t* pos, void* key, size_t hash,
   size_t index = ((hash & mask) + p_length) & mask;
   void* elem;
   size_t elem_hash;
+  size_t ib_index;
+  size_t ib_offset;
 
   for(size_t i = 0; i <= mask; i++)
   {
+    ib_index = index/HASHMAP_BITMAP_TYPE_SIZE;
+    ib_offset = index%HASHMAP_BITMAP_TYPE_SIZE;
     elem = map->buckets[index].ptr;
     elem_hash = map->buckets[index].hash;
 
-    if(elem == NULL)
+    // if bucket is empty
+    if((map->item_bitmap[ib_index] & ((bitmap_t)1 << ib_offset)) == 0)
```

Comment:
The goal is to eliminate the cache line fetch when the bucket is empty. Whether it will help or not... we'll see :)